### PR TITLE
Add a StringToNumberConverter

### PR DESCRIPTION
**Pull request by gselzer**
This PR adds a `StringToNumberConverter`, a dedicated `Converter` that converts from `String`s to:
* numerical primitives
* `Number`
* `Number`'s subclasses

This is useful for imagej/imagej-ops#644

### CODE REVIEW

File: src/test/java/org/scijava/convert/StringToNumberConverterTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Tests {@link StringToNumberConverter}
+ *
+ * @author Gabriel Selzer
+ */
+public class StringToNumberConverterTest {
```
Comment (ctrueden) [inline src/test/java/org/scijava/convert/StringToNumberConverterTest.java:13]: Needs tests that non-numerically-formatted strings successfully report `false` for `canConvert`.

File: src/main/java/org/scijava/convert/StringToNumberConverter.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	@Override
+	public Class<String> getInputType() {
+		return String.class;
+	}
+
+	@Override
+	public boolean canConvert(Object src, Class<?> dest) {
+		if (!Types.isAssignable(src.getClass(), String.class)) return false;
+		return Types.isAssignable(Types.box(dest), Number.class);
```
Comment (ctrueden) [inline src/main/java/org/scijava/convert/StringToNumberConverter.java:54]: Needs to check that the string is actually a valid numeric format. The best way to do that is, sadly, to actually do the conversion. But that's OK.

**Review by ctrueden — CHANGES_REQUESTED** [review]
As written, this will cause the convert service to erroneously report strings like `"Hello"` as convertible to `Number` subclasses. We need to be selective in which strings we allow here, no? In the same way we are being selective e.g. with unwrapping `ROITree` objects?

**Comment by gselzer** [comment]
Thanks for the feedback @ctrueden!

Both your concerns should be fixed in 5c160d2

File: src/main/java/org/scijava/convert/StringToNumberConverter.java [hunk 3]
```diff
@@ -51,6 +50,22 @@ public Class<String> getInputType() {
 	@Override
 	public boolean canConvert(Object src, Class<?> dest) {
 		if (!Types.isAssignable(src.getClass(), String.class)) return false;
-		return Types.isAssignable(Types.box(dest), Number.class);
+		// The only way to know if the conversion is valid is to actually do it.
+		try {
+			String srcString = (String) src;
+			sane(dest).getConstructor(String.class).newInstance(srcString);
```
Comment (ctrueden) [inline src/main/java/org/scijava/convert/StringToNumberConverter.java:56]: This reflection, while slick, concerns me w.r.t. the Java 11+ restrictions on reflection. But it's fine for now, we can always fix later as needed.

Comment (gselzer) [reply src/main/java/org/scijava/convert/StringToNumberConverter.java:56]: Yes, you are right. `DefaultConverter` does the same thing, though, so I think you are right to fix these later.

**Pull request merged**